docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,505 @@
+
+These changes update the data row declarations for both tests. The following code shows these tests with all data row attributes:
+
+[!code-csharp[The finished tests](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#FinishedTests "The finished tests for the make const analyzer")]
+
+Fortunately, all of the above bugs can be addressed using the same techniques that you just learned.
+
+To fix the first bug, first open **DiagnosticAnalyzer.cs** and locate the foreach loop where each of the local declaration's initializers are checked to ensure that they're assigned with constant values. Immediately _before_ the first foreach loop, call `context.SemanicModel.GetTypeInfo()` to retrieve detailed information about the declared type of the local declaration:
+
+```csharp
+var variableTypeName = localDeclaration.Declaration.Type;
+var variableType = context.SemanticModel.GetTypeInfo(variableTypeName).ConvertedType;
+```
+
+Then, inside your `foreach` loop, check each initializer to make sure it's convertible to the variable type. Add the following check after ensuring that the initializer is a constant:
+
+```csharp
+// Ensure that the initializer value can be converted to the type of the
+// local declaration without a user-defined conversion.
+var conversion = context.SemanticModel.ClassifyConversion(initializer.Value, variableType);
+if (!conversion.Exists || conversion.IsUserDefined)
+{
+    return;
+}
+```
+
+The next change builds upon the last one. Before the closing curly brace of the first foreach loop, add the following code to check the type of the local declaration when the constant is a string or null.
+
+```csharp
+// Special cases:
+//  * If the constant value is a string, the type of the local declaration
+//    must be System.String.
+//  * If the constant value is null, the type of the local declaration must
+//    be a reference type.
+if (constantValue.Value is string)
+{
+    if (variableType.SpecialType != SpecialType.System_String)
+    {
+        return;
+    }
+}
+else if (variableType.IsReferenceType && constantValue.Value != null)
+{
+    return;
+}
+```
+
+You must write a bit more code in your code fix provider to replace the var' keyword with the correct type name. Return to **CodeFixProvider.cs**. The code you'll add does the following steps:
+
+* Check if the declaration is a `var` declaration, and if it is:
+* Create a new type for the inferred type.
+* Make sure the type declaration is not an alias. If so, it is legal to declare `const var`.
+* Make sure that `var` isn't a type name in this program. (If so, `const var` is legal).
+* Simplify the full type name
+
+That sounds like a lot of code. It's not. Add the following code immediately after the initialization of `newModifiers`:
+
+[!code-csharp[Replace Var designations](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConstCodeFixProvider.cs#ReplaceVar "Replace a var designation with the explicit type")]
```

Comment:
This code created a conflict with the `newLocal` variable. It already exists and is slightly different. The steps on where to paste this code did not specify that the existing `newLocal` variable needed to be deleted.